dartagnan/src/main/java/com/dat3m/dartagnan/program/processing/NaiveDevirtualisation.java

```diff
@@ -131,61 +136,108 @@ private void devirtualise(Function function, Map<Function, IValue> func2AddressM
+            // Construct call table
+            for (Function possibleTarget : possibleTargets) {
+                final IValue targetAddress = func2AddressMap.get(possibleTarget);
+                final Label caseLabel = EventFactory.newLabel(String.format("__Ldevirt_%s#%s", targetAddress.getValue(), devirtCounter));
+                final CondJump caseJump = EventFactory.newJump(expressions.makeEQ(funcPtr, targetAddress), caseLabel);
+                caseLabels.add(caseLabel);
+                caseJumps.add(caseJump);
+            }
+
+            final Event noMatch = EventFactory.newAssert(expressions.makeFalse(), "Invalid function pointer");
+            final Label endLabel = EventFactory.newLabel(String.format("__Ldevirt_end#%s", devirtCounter));
+
+            final List<Event> callReplacement = new ArrayList<>();
+            callReplacement.add(EventFactory.newStringAnnotation("=== Devirtualized call ==="));
+            callReplacement.addAll(caseJumps);
+            callReplacement.add(noMatch);
+            for (int i = 0; i < caseLabels.size(); i++) {
+                callReplacement.add(caseLabels.get(i));
+                callReplacement.add(devirtualiseCall(call, possibleTargets.get(i)));
+                callReplacement.add(EventFactory.newGoto(endLabel));
+            }
+            callReplacement.add(endLabel);
+            callReplacement.add(EventFactory.newStringAnnotation("=== End of devirtualized call ==="));
+
+            call.replaceBy(callReplacement);
+            callReplacement.forEach(e -> e.copyAllMetadataFrom(call));
 
-                call.replaceBy(callReplacement);
-                callReplacement.forEach(e -> e.copyAllMetadataFrom(call));
+            devirtCounter++;
+        }
+    }
+
+    private boolean needsDevirtualization(FunctionCall call) {
+        return !call.isDirectCall() ||
+                (call.getCalledFunction().getIntrinsicInfo() == Intrinsics.Info.P_THREAD_CREATE
+                && !(call.getArguments().get(2) instanceof Function));
+    }
+
+    private List<Function> getPossibleTargets(FunctionCall call, Map<Function, IValue> func2AddressMap) {
+        final List<Function> possibleTargets;
+        if (!call.isDirectCall()) {
+            possibleTargets = func2AddressMap.keySet().stream()
+                    .filter(f -> f.getFunctionType() == call.getCallType()).collect(Collectors.toList());
+        } else if (call.getCalledFunction().getIntrinsicInfo() == Intrinsics.Info.P_THREAD_CREATE) {
+            final TypeFactory types = TypeFactory.getInstance();
+            final Type ptrType = types.getPointerType();
+            final Type threadType = types.getFunctionType(ptrType, List.of(ptrType));
+            possibleTargets = func2AddressMap.keySet().stream()
+                    .filter(f -> f.getFunctionType() == threadType).collect(Collectors.toList());
+        } else {
+            possibleTargets = List.of();
+            assert false;
+        }
```

Comment:
I don't like this kind of asserts: if they fail, this tells nothing to the user. If we expect that after passing the first `if` we will certainly encounter some given functions (now only pthread_create, but possible more in the future), this assumption should be the one being made in the assert.

Also, the empty list of possible target is useless since the assert would terminate the execution

Reply:
Unfortuntely, I have to assign `possibleTargets` for otherwise the compiler complaints.

I disagree with having an `assert xyz` instead of the last else-case. A reason is that the assert has to change whenever a new function is added, i.e., it is badly extensible. And this code is expected to get extended later on.

The code should not be reachable in principle, no matter what code it runs on.
More precisely, if that location is ever reached, it does not mean we have some bad input or some not-yet supported input, but rather the logic of this class is broken.
I can change it to `Verify.verify` which is the closest to Java's assertions, though it's a little less fitting.

Reply:
The branches in `getPossibleTargets` should cover the same cases as `needsDevirtualization`, right? If so, I would rather use `Precondition.checkArguments` than having the assertion

Reply:
Yes, in all cases where `assert false` is placed, the assumption is that the checks cover all cases of `needsDevirtualization` (if they don't, there should be an error).
Not sure how you wanna use `Precondition.checkArguments` here, cause if you think about `Precondition.checkArguments(needsDevirtualiatuon(...))` then this is precisely what is not smart to do, because it passes whenever `needsDevirtualization` gets extended.

The only right thing to do is to throw an error in the else-case I think.